### PR TITLE
Add cranking taper duration multiplier based on CLT

### LINKED ISSUE #369: Add cranking IAC taper duration multiplier
I was wondering, would it be useful to have an `After cranking IAC taper duration multiplier override table`?

I enabled the `Override cranking IAC CLT multiplier` to change the IAC position based on the temperature, which really helps when the engine gets warm and doesn't get too much air when cranking.

But the `After cranking IAC taper duration` is always a fixed (in my case, 200 cycles) number. When an engine is warm, the IAC could taper down sooner because we don't need to hold the RPMs as high and for that long.

### PR DESCRIPTION
**Pull request by Krakert**
Fixes #369

### CODE REVIEW

File: firmware/tunerstudio/rusefi.input [hunk 1]
```diff
@@ -3823,6 +3835,7 @@ cmd_set_engine_type_default					= "@@TS_IO_TEST_COMMAND_char@@@@ts_command_e_TS_
 		field = "Cranking base IAC position",			crankingIACposition
 		field = "After cranking IAC taper duration",afterCrankingIACtaperDuration
```
Comment (mck1117) [inline firmware/tunerstudio/rusefi.input:3833]: add a condition here to disable this setting when the new switch is turned on

File: firmware/integration/rusefi_config.txt [hunk 2]
```diff
@@ -981,6 +981,7 @@ bit skippedWheelOnCam,"On camshaft","On crankshaft";Where is your primary skippe
 	bit complexWallModel,"Advanced (tables)","Basic (constants)";Should we use tables to vary tau/beta based on CLT/MAP, or just with fixed values?
 	bit alwaysInstantRpm
 	bit isMapAveragingEnabled
+	bit overrideCrankingTaperDurationSetting; If enabled, use separate temperature multiplier table for cranking taper duration.
```
Comment (mck1117) [inline firmware/integration/rusefi_config.txt:984]: I'm not sure `override` is the right language here - maybe `useCrankingIdleTaperTable` or something would be clearer?

**Comment by mck1117** [comment]
I'd love to see a unit test for the new behavior :)

File: firmware/tunerstudio/rusefi.input (hunk 1 above)
Comment (Krakert) [reply firmware/tunerstudio/rusefi.input:3833]: We still use the `afterCrankingIACtaperDuration` or do you mean something else?

File: firmware/integration/rusefi_config.txt (hunk 2 above)
Comment (Krakert) [reply firmware/integration/rusefi_config.txt:984]: Will change, just looked add `overrideCrankingIacSetting`

File: firmware/controllers/actuators/idle_thread.cpp [hunk 3]
```diff
@@ -79,8 +79,13 @@ IIdleController::Phase IdleController::determinePhase(int rpm, int targetRpm, Se
...
 
-float IdleController::getCrankingTaperFraction() const {
-	return (float)engine->rpmCalculator.getRevolutionCounterSinceStart() / engineConfiguration->afterCrankingIACtaperDuration;
+float IdleController::getCrankingTaperFraction(float clt) const {
+    if (engineConfiguration->overrideCrankingTaperDurationSetting) {
+        return (float) engine->rpmCalculator.getRevolutionCounterSinceStart() /
+               (engineConfiguration->afterCrankingIACtaperDuration *
+                interpolate2d(clt, config->cltCrankingTaperCorrBins, config->cltCrankingTaperCorr));
+    }
+    return (float)engine->rpmCalculator.getRevolutionCounterSinceStart() / engineConfiguration->afterCrankingIACtaperDuration;
```
Comment (mck1117) [inline firmware/controllers/actuators/idle_thread.cpp:88]: I would avoid duplicating this code if you can - it's not obvious to the reader why these two cases are different when they look so similar on first glance.

prefer something like

```c++
float x = config.taperDuration;

if (new_behavior) {
    x *= table_lookup();
}

return getRevCounter() / x;
```

Comment (mck1117) [reply firmware/controllers/actuators/idle_thread.cpp:88]: that style of writing it makes it very clear which parts are the same, and which parts are different when the setting is enabled

File: firmware/integration/rusefi_config.txt [hunk 4]
```diff
@@ -1556,6 +1557,9 @@ uint8_t[PEDAL_TO_TPS_SIZE] autoscale pedalToTpsRpmBins;;"RPM", 100, 0, 0, 25000,
 float[CLT_CRANKING_CURVE_SIZE] cltCrankingCorrBins;CLT-based cranking position multiplier for simple manual idle controller;"C", 1, 0, -100, 250, 2
 float[CLT_CRANKING_CURVE_SIZE] cltCrankingCorr    ;CLT-based cranking position multiplier for simple manual idle controller;"%", 1, 0, 0, 500, 2
 
+float[CLT_CRANKING_CURVE_SIZE] cltCrankingTaperCorrBins;CLT-based taper duration multiplier for simple manual idle controller;"C", 1, 0, -100, 250, 2
+float[CLT_CRANKING_CURVE_SIZE] cltCrankingTaperCorr    ;CLT-based taper duration multiplier for simple manual idle controller;"%", 1, 0, 0, 500, 2
```
Comment (mck1117) [inline firmware/integration/rusefi_config.txt:1561]: 1. is 500 a reasonable maximum value for this parameter? maybe 5 is safer?
2. using float here is a waste of storage, maybe int8_t for temp, and uint8_t with scale 0.02 for multiplier would be better?

Comment (Krakert) [reply firmware/integration/rusefi_config.txt:1561]: Maybe a good idea to also make these change this for the `cltCrankingCorr`

Comment (mck1117) [reply firmware/integration/rusefi_config.txt:1561]: yes, finding other tables we can compact would be a great follow-on task #371

**Comment by Krakert** [comment]
> I'd love to see a unit test for the new behavior :)

Added a unit test

**Review by mck1117 — APPROVED** [review]

File: firmware/tunerstudio/rusefi.input (hunk 1 above)
Comment (mck1117) [reply firmware/tunerstudio/rusefi.input:3833]: oop, yeah, it's a multiplier on that duration, not just the duration. lgtm.

**Pull request merged**